### PR TITLE
Improve picking usage

### PR DESCRIPTION
**Pull request by nmellado**
* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Be aware that the PR request cannot be accepted if it doesn't pass the Continuous Integration tests.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
API improvements + doc

* **What is the current behavior?** (You can also link to an open issue here)
PickingResult store separately the indices of picked faces, vertices and edges.

* **What is the new behavior (if this is a feature change)?**
Group face, vertex and edge id in a tuple to ease processing + add duplicate removal tools.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
Change of API.

* **Other information**:

### CODE REVIEW

**Comment by nmellado** [comment]
Ready for review.

File: tests/ExampleApps/PickingDemoApp/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+project(PickingDemoApp VERSION 0.0.1)
+
+if (NOT CMAKE_BUILD_TYPE)
+    set(CMAKE_BUILD_TYPE Release)
+endif ()
+
+# Set default install location to installed-<Compiler_ID> folder in build dir
+# we do not want to install to /usr by default
+if (CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
```
Comment (dlyr) [inline tests/ExampleApps/PickingDemoApp/CMakeLists.txt:20]: This change the main `CMAKE_INSTALL_PREFIX` on a fresh cmake configuration
(the value `CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT` do not change its value when `set(CMAKE_INSTALL_PREFIX` ...) is called in main `CMakeLists.txt`

Comment (dlyr) [reply tests/ExampleApps/PickingDemoApp/CMakeLists.txt:20]: Resolved by fffaeb9

File: src/Engine/Rendering/Renderer.hpp [hunk 2]
```diff
@@ -94,19 +94,55 @@ class RA_ENGINE_API Renderer
...
-        std::vector<int> m_elementIdx; // Idx of the element, i.e. triangle for mesh, edge for lines
-                                       // and -1 for points
-        std::vector<int> m_edgeIdx;    // Idx of the opposite vertex in the triangle if mesh
-        // Note: There is exactly one triplet for each querried pixel (thus there can be doublons
-        // coming from pixels). Note: Beware that the same mesh vertex would also be picked for each
-        // of its adjacent triangles.
+        /// Picking mode of the query
+        PickingMode m_mode {Engine::Rendering::Renderer::RO};
+        /// Idx of the picked RO
+        Core::Utils::Index m_roIdx {Core::Utils::Index::Invalid()};
```
Comment (dlyr) [inline src/Engine/Rendering/Renderer.hpp:116]: Is this field public ?

File: src/Engine/Rendering/Renderer.hpp [hunk 3]
```diff
@@ -94,19 +94,55 @@ class RA_ENGINE_API Renderer
...
+
+      private:
```
Comment (dlyr) [inline src/Engine/Rendering/Renderer.hpp:118]: Usually we have first public section, then private

File: src/Engine/Rendering/Renderer.hpp [hunk 4]
```diff
@@ -94,19 +94,55 @@ class RA_ENGINE_API Renderer
...
+        /// Query result, stored as: [vertexId, elementId, edgeId]
+        /// \see removeDuplicatedIndices()
+        ///
+        /// \note Set as mutable to be able to call removeDuplicatedIndices() in const context.
+        mutable std::vector<std::tuple<int, int, int>> m_indices;
+
+      public:
+        /// Read access to the collected ids
+        inline const std::vector<std::tuple<int, int, int>>& indices() const { return m_indices; }
```
Comment (dlyr) [inline src/Engine/Rendering/Renderer.hpp:127]:
```suggestion
        inline const std::vector<std::tuple<int, int, int>>& getIndices() const { return m_indices; }
```

File: src/Engine/Rendering/Renderer.hpp [hunk 5]
```diff
@@ -94,19 +94,55 @@ class RA_ENGINE_API Renderer
     };
 
     /**
-     * Picking result
+     * \brief Result of a PickingQuery
+     *
+     * For each selected pixel, store the following indices:
```
Comment (dlyr) [inline src/Engine/Rendering/Renderer.hpp:99]:
```suggestion
     * For each selected pixel, store the following indices: `<elementIdx, vertexIdx, edgeIdx>`
```

File: src/Engine/Rendering/Renderer.hpp [hunk 6]
```diff
@@ -94,19 +94,55 @@ class RA_ENGINE_API Renderer
...
+     *  - `elementId`: Id of the selected face or edge (for mesh and line mesh respectively). Set to
```
Comment (dlyr) [inline src/Engine/Rendering/Renderer.hpp:100]:
```suggestion
     *  - `elementIdx`: Index of the selected face or edge (for mesh and line mesh respectively). Set to
```

File: src/Engine/Rendering/Renderer.hpp [hunk 7]
```diff
@@ -94,19 +94,55 @@ class RA_ENGINE_API Renderer
...
+     *  -1 for point clouds.
+     *  - `vertexId`: On meshes/lines: id of the closest vertex on the selected face/edge. On
```
Comment (dlyr) [inline src/Engine/Rendering/Renderer.hpp:102]:
```suggestion
     *  - `vertexIdx`: On meshes/lines: index of the closest vertex on the selected face/edge. On
```

File: src/Engine/Rendering/Renderer.hpp [hunk 8]
```diff
@@ -94,19 +94,55 @@ class RA_ENGINE_API Renderer
...
+     *  point-clouds: index of the selected vertex
+     *  - `edgeId`: On meshes/lines: id of the closest edge on the selected face/edge. In meshes,
```
Comment (dlyr) [inline src/Engine/Rendering/Renderer.hpp:104]:
```suggestion
     *  - `edgeIdx`: On meshes/lines: index of the closest edge on the selected face/edge. In meshes,
```

File: src/Engine/Rendering/Renderer.hpp [hunk 9]
```diff
@@ -94,19 +94,55 @@ class RA_ENGINE_API Renderer
...
+     *  `edgeid` is defined using the index of the opposite vertex. Set to -1 for point clouds.
```
Comment (dlyr) [inline src/Engine/Rendering/Renderer.hpp:105]:
```suggestion
     *  `edgeIdx` is defined using the index of the opposite vertex. Set to -1 for point clouds.
```

File: tests/ExampleApps/PickingDemoApp/main.cpp [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    // selection (<shift+whell modify the radius) This Radius allow to select things for all pixels
+    // within the circle (with the "T", "V" or "E" key pressed while clicking) For each pixel i
+    // within the circle, whe have, for the render object below the mouse ( pickingResult.m_roIdx ):
+    //      pickResults.m_elementIdx : the nearest triangle
+    //      pickResults.m_vertexIdx  : the nearest vertex in the triangle(0, 1 or 2)
+    //      pickResults.m_edgeIdx    : the nearest edge in the triangle
+    void handlePicking( const Ra::Engine::Rendering::Renderer::PickingResult& pickingResult ) {
+        if ( m_isTracking )
+        {
+            std::cout << "Circle selection mode : " << std::endl;
```
Comment (dlyr) [inline tests/ExampleApps/PickingDemoApp/main.cpp:62]: Is there any reason to not use Radium's LOG instead ?

File: tests/ExampleApps/PickingDemoApp/main.cpp [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    //! [Initializing the application]
+    // The customization functor is called here
+    app.initialize( DemoWindowFactory() );
+    //! [Initializing the application]
+
+    if ( argc < 2 )
+    {
+        // Create a scene when called without input file
+        addDummyScene( app );
+    }
```
Comment (dlyr) [inline tests/ExampleApps/PickingDemoApp/main.cpp:150]: Is there any else ? I have tried if something magic here, but giving an arg to the app do not display anything.

**Review by dlyr — COMMENTED** [review]
I'm fine with the overall PR, some comments and questions.
the suggestion `indices() -> getIndices()` need to be propagated on the rest of the code (if OK I can do it)

File: src/Engine/Rendering/Renderer.hpp (hunk 2 above)
Comment (nmellado) [reply src/Engine/Rendering/Renderer.hpp:116]: The struct had only public fields. I moved one to private, but true, we could move to a class and hide them all.

File: src/Engine/Rendering/Renderer.hpp (hunk 3 above)
Comment (nmellado) [reply src/Engine/Rendering/Renderer.hpp:118]: Agree, see my previous comment.

File: src/Engine/Rendering/Renderer.hpp (hunk 4 above)
Comment (nmellado) [reply src/Engine/Rendering/Renderer.hpp:127]: Ok fine

File: tests/ExampleApps/PickingDemoApp/main.cpp (hunk 10 above)
Comment (nmellado) [reply tests/ExampleApps/PickingDemoApp/main.cpp:62]: No, we should use Log

File: tests/ExampleApps/PickingDemoApp/main.cpp (hunk 11 above)
Comment (nmellado) [reply tests/ExampleApps/PickingDemoApp/main.cpp:150]: See base application options. You need -f to load a scene. My check is a bit hacky, true.

**Review by dlyr — APPROVED** [review]

File: src/Engine/Rendering/Renderer.cpp [hunk 12]
```diff
@@ -193,7 +193,7 @@ Renderer::PickingResult Renderer::doPickingNow( const PickingQuery& query,
     if ( query.m_screenCoords.x() < 0 || query.m_screenCoords.x() > m_width - 1 ||
          query.m_screenCoords.y() < 0 || query.m_screenCoords.y() > m_height - 1 )
     {
-        result.m_roIdx = -1;
+        result.setRoIdx( Ra::Core::Utils::Index::Invalid() );
```
Comment (nmellado) [inline src/Engine/Rendering/Renderer.cpp:196]:
```suggestion
        result.clear(); // reset to empty state
```

Comment (dlyr) [reply src/Engine/Rendering/Renderer.cpp:196]: Actually, we can simply return result, or {} since nothing is set from line 190 (decl) to here.

Comment (nmellado) [reply src/Engine/Rendering/Renderer.cpp:196]: Ok return {} sounds good.
There is another line in the code that looks the same.

**Comment by nmellado** [comment]
Ready to review.

File: CMakeLists.txt [hunk 13]
```diff
@@ -110,6 +110,9 @@ if (CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
     set(CMAKE_INSTALL_PREFIX "${RADIUM_BUNDLE_DIRECTORY}" CACHE PATH
         "Install path prefix, prepended onto install directories." FORCE)
     messageInfo("Set install prefix to ${CMAKE_INSTALL_PREFIX}")
+    # prevent subsequent modiifcation of CMAKE_INSTALL_PREFIX
```
Comment (nmellado) [inline CMakeLists.txt:113]:
```suggestion
    # prevent subsequent modification of CMAKE_INSTALL_PREFIX
```

**Comment by nmellado** [comment]
Can someone review this PR ?

**Review by dlyr — APPROVED** [review]

**Pull request merged**